### PR TITLE
feat(rule ticket): add mail headers as criteria

### PR DESCRIPTION
**Pull request by stonebuzz**
Add new criterias to rule ticket engine (like ```x-priority```)

- from
- subject
- reply-to
- to
- in-reply-to

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !24256

### CODE REVIEW

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

File: tests/functionnal/RuleTicket.php [hunk 1]
```diff
@@ -2220,16 +2220,51 @@ public function testAppendAppliance()
         ))->isEqualTo(2);
     }
 
-    public function testHeaderCriteria(): void
+    protected function testMailHeaderCriteriaProvider()
     {
+        return [
+            [
+                "header"  => '_x-priority',
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2227]:
```suggestion
                "header"  => 'x-priority',
```

File: tests/functionnal/RuleTicket.php [hunk 2]
```diff
@@ -2240,9 +2275,9 @@ public function testHeaderCriteria(): void
 
         $crit_id = $rulecrit->add($crit_input = [
             'rules_id'  => $ruletid,
-            'criteria'  => '_x-priority',
+            'criteria'  => $header,
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2278]:
```suggestion
            'criteria'  => '_' . $header,
```

File: tests/functionnal/RuleTicket.php [hunk 3]
```diff
@@ -2255,26 +2290,26 @@ public function testHeaderCriteria(): void
...
+
         $ticket = new \Ticket();
         $tickets_id = $ticket->add([
-            'name'              => 'test x-priority header',
-            'content'           => 'test x-priority header',
+            'name'              => 'test ' . $header . ' header',
+            'content'           => 'test ' . $header . ' header',
             '_head'             => [
-                'x-priority' => '1'
+                substr($header, 1) => $pattern
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2299]:
```suggestion
                $header => $pattern
```

File: tests/functionnal/RuleTicket.php [hunk 4]
```diff
@@ -2255,26 +2290,26 @@ public function testHeaderCriteria(): void
...
-        // Retest ticket with different x-priority
+        // Retest ticket with different header
         $tickets_id = $ticket->add([
-            'name'              => 'test x-priority header',
-            'content'           => 'test x-priority header',
+            'name'              => 'test ' . $header . ' header',
+            'content'           => 'test ' . $header . ' header',
             '_head'             => [
-                'x-priority' => '2'
+                substr($header, 1) => 'header_foo_bar'
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2312]:
```suggestion
                $header => 'header_foo_bar'
```

File: tests/functionnal/RuleTicket.php [hunk 5]
```diff
@@ -2220,16 +2220,51 @@ public function testAppendAppliance()
...
+    ) {
         $this->login();
 
         $ruleticket = new \RuleTicket();
         $rulecrit   = new \RuleCriteria();
         $ruleaction = new \RuleAction();
 
         $ruletid = $ruleticket->add($ruletinput = [
-            'name'         => 'test x-priority',
+            'name'         => 'test from',
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2267]:
```suggestion
            'name'         => 'test ' . $header,
```

File: tests/functionnal/RuleTicket.php [hunk 6]
```diff
@@ -2220,16 +2220,51 @@ public function testAppendAppliance()
...
-    public function testHeaderCriteria(): void
+    protected function testMailHeaderCriteriaProvider()
     {
+        return [
+            [
+                "header"  => '_x-priority',
+                "pattern" => 'pattern_1',
+            ],
+            [
+                "header"  => '_from',
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2231]:
```suggestion
                "header"  => 'from',
```

File: tests/functionnal/RuleTicket.php [hunk 7]
```diff
@@ -2220,16 +2220,51 @@ public function testAppendAppliance()
...
+                "pattern" => 'pattern_from',
+            ],
+            [
+                "header"  => '_to',
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2235]:
```suggestion
                "header"  => 'to',
```

File: tests/functionnal/RuleTicket.php [hunk 8]
```diff
@@ -2220,16 +2220,51 @@ public function testAppendAppliance()
...
+                "pattern" => 'pattern_to',
+            ],
+            [
+                "header"  => '_reply-to',
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2239]:
```suggestion
                "header"  => 'reply-to',
```

File: tests/functionnal/RuleTicket.php [hunk 9]
```diff
@@ -2220,16 +2220,51 @@ public function testAppendAppliance()
...
+                "pattern" => 'pattern_reply-to',
+            ],
+            [
+                "header"  => '_in-reply-to',
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2243]:
```suggestion
                "header"  => 'in-reply-to',
```

File: tests/functionnal/RuleTicket.php [hunk 10]
```diff
@@ -2220,16 +2220,51 @@ public function testAppendAppliance()
...
+                "pattern" => 'pattern_in-reply-to',
+            ],
+            [
+                "header"  => '_subject',
```
Comment (cedric-anne) [inline tests/functionnal/RuleTicket.php:2247]:
```suggestion
                "header"  => 'subject',
```

**Review by cedric-anne — COMMENTED** [review]
It would be more easy to read the code with the `_` only added when required.

**Review by orthagh — APPROVED** [review]

**Pull request merged**